dhl_sdk/_input_processing.py

```diff
@@ -0,0 +1,718 @@
+    InvalidSpectraException,
+    InvalidInputsException,
+    InvalidStepsException,
+    InvalidTimestampsException,
+)
+from dhl_sdk._spectra_utils import (
+    SpectraData,
+    SpectraModel,
+    _validate_spectra_format,
+    _convert_to_request,
+)
+
+
+class Group(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def code(self) -> str: ...
+
+
+class Variable(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def group(self) -> Group: ...
+
+    @property
+    def code(self) -> str: ...
+
+    @property
+    def id(self) -> str: ...
+
+    def matches_key(self, key: str) -> bool: ...
+
+
+class Dataset(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def variables(self) -> list[Variable]: ...
+
+
+class Model(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def dataset(self) -> Dataset: ...
+
+
+# Input Preprocessors (Validation and Formatting)
+class Preprocessor(ABC):
+    """Interface for preprocessing"""
+
+    @abstractmethod
+    def validate(self) -> bool:
+        """Validate the input data"""
+
+    @abstractmethod
+    def format(self) -> list[dict]:
+        """Format the input data to a list of JSON requests"""
+
+
+@dataclass
+class SpectraPreprocessor(Preprocessor):
+    """Implementation of the preprocessor for Spectrum models"""
+
+    spectra: SpectraData
+    inputs: Optional[dict]
+    model: SpectraModel
+
+    def validate(self) -> bool:
+        """
+        This function validates and makes the necessary formating of the spectra
+        and inputs used for prediction. It performs the following validations:
+
+        - Validates if spectra is empty
+        - Validates if the number of wavelengths in spectra matches
+            the number of wavelengths in the model
+        - Validates if the spectra contains None values
+        - Validates if the number of inputs matches model inputs
+        - Validates if the number of values in each input matches the number of spectra
+
+        Parameters
+        ----------
+        spectra : Union[list[list[float]], np.ndarray]
+            The spectra to be validated
+        model : Model
+            The model to use for prediction.
+        inputs : dict, optional
+            A dictionary of input variables and their values, by default None
+
+        Returns
+        -------
+        list[list[float], dict
+            If all validations pass, return the formatted spectra and inputs
+            as a tuple.
+
+        Raises
+        ------
+        InvalidSpectraException
+            Exception raised when Spectra is not valid for prediction.
+        InvalidInputsException
+            Exception raised when Inputs is not valid for prediction.
```

Comment:
i know that Spectra and Inputs are individual (singular) variables, but i still think we should, when writing, deal with them as plural entities in text, so:
"Exception raised when Spectra are not valid for prediction"
"Exception raised when Inputs are not valid for prediction"

but up to you

edit: below you use plural, which imo reads better, so im much in favor of using plural